Add example blueprint for CSI snapshot functions (#1184)
kanisterio/kanister: examples/stable/mysql-csi-snapshot/README.md

# Backup and restore MySQL DB using CSI VolumeSnapshot

VolumeSnapshots provide Kubernetes users with a standardized way to copy a volume's contents at a particular point in time without creating an entirely new volume. This functionality enables, for example, database administrators to backup databases before performing edit or delete modifications.

## Introduction

> This document explains how Kanister leverages the use of CSI VolumeSnapshots to capture and restore MySQL database.

Reviewer: As discussed offline, let's use a non-DB application for this example. Generally, I assume that the blueprint would be used directly on PVCs but if we want to talk about an application, let's use timelogger.

> # MySQL DB using CSI VolumeSnapshots

Reviewer 2: Title doesnt specify that we are doing backup/restore

Author: Should we set the title as "Backup and restore MySQL DB using CSI VolumeSnapshot"?

VolumeSnapshots provide Kubernetes users with a standardized way to copy a volume's contents at a particular point in time without creating an entirely new volume. This functionality enables, for example, database administrators to backup databases before performing edit or delete modifications.

## Introduction
This document explains how Kanister leverages the use of CSI VolumeSnapshots to capture and restore MySQL database.

## Prerequisites

- Helm 3 installed
- Kubernetes 1.16+ with Beta APIs enabled
- Kanister controller version 0.71.0 installed in your cluster, let's assume in Namespace `kanister`
- Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)
- VolumeSnapshots should be supported by CSI Driver and respective CRDs should be available

## Install MySQL

Install the MySQL database using the bitnami chart with the release name `mysql-release`.

```bash
# Add bitnami in your local chart repository
$ helm repo add bitnami https://charts.bitnami.com/bitnami

# Update your local chart repository
$ helm repo update

> # Install the MySQL database (Helm Version 3)

Reviewer: ```suggestion
# Install the MySQL database
```

$ kubectl create namespace mysql
$ helm install mysql-release bitnami/mysql --namespace mysql \
    --set auth.rootPassword='asd#45@mysqlEXAMPLE' \
    --set architecture="standalone"
```

Above command deploys a MySQL instance in the `mysql` namespace.

> To retrieve your root password run the following command.

Reviewer: ```suggestion
Retrieve the root password with the following command.
```

```bash
$ kubectl get secret mysql-release --namespace mysql -o jsonpath="{.data.mysql-root-password}" | base64 --decode`
```

> > **Tip**: List all releases using `helm list --all-namespaces`, using Helm Version 3.

Reviewer: ```suggestion
> **Tip**: List all releases using `helm list --all-namespaces`.
```
since we already have helm3 as a pre-req

```

From inside the shell, use the mysql CLI to insert some data into the test database.

```bash
$ mysql --user=root --password=asd#45@mysqlEXAMPLE

mysql> CREATE DATABASE test;
Query OK, 1 row affected (0.00 sec)

mysql> USE test;
Database changed

# Create "employees" table
mysql> CREATE TABLE employees (name VARCHAR(20), manager VARCHAR(20), address VARCHAR(200), sex CHAR(1), birth DATE, department VARCHAR(20));
Query OK, 0 rows affected (0.02 sec)

# Insert row to the table
mysql> INSERT INTO employees VALUES ('Max','Tom','Naperville, Illinois, Chicago','m','1999-03-30','Engineering');
Query OK, 1 row affected (0.01 sec)

# View data in "employees" table
mysql> SELECT * FROM employees;
+------+---------+-------------------------------+------+------------+-------------+
| name | manager | address                       | sex  | birth      | department  |
+------+---------+-------------------------------+------+------------+-------------+
| Max  | Tom     | Naperville, Illinois, Chicago | m    | 1999-03-30 | Engineering |
+------+---------+-------------------------------+------+------------+-------------+
1 row in set (0.00 sec)
```

## Backup MySQL DB

### Create Blueprint

> Create Blueprint in the same namespace as the Kanister controller.

Reviewer: ```suggestion
Create the Blueprint in the `kanister` namespace.
```
Some instructions directly say this

> > **Note**:  We have set up the Kubernetes cluster on DigitalOcean. Hence, `snapshotClass` and `storageClass` in the `./mysql-csi-snapshot-blueprint.yaml` file is set to `do-block-storage`. Please correct these arguments as per your cluster setup before creating the blueprint.

Reviewer: ```suggestion
> **Note**:  This example uses a Kubernetes cluster on DigitalOcean. Therefore the `snapshotClass` and `storageClass` in the `./mysql-csi-snapshot-blueprint.yaml` file are set to `do-block-storage`. Change the arguments appropriately before creating the blueprint.
```

Create Blueprint in the same namespace as the Kanister controller.

> **Note**: We used a Kubernetes cluster on DigitalOcean. Hence, snapshotClass and storageClass in the ./mysql-csi-snapshot-bp.yaml file is set to `do-block-storage`. Please correct these arguments as per your cluster setup. You can make use of `kubectl patch` or `kubectl edit` commands to make these changes.

```bash

> $ kubectl create -f ./mysql-csi-snapshot-bp.yaml -n kanister

Reviewer 3: file is named mysql-csi-snapshot-**blueprint**.yaml

Author: Oh. My bad. I missed that.

```

## Backup the application data

> Take a backup of the MySQL data using the backup ActionSet from above blueprint. Create an ActionSet in the same namespace as the controller. An easy way to do this is by using ``kanctl``.

Reviewer 2: ```suggestion
Take a backup of the MySQL data using the backup ActionSet from above blueprint. Create an ActionSet in the `kanister` namespace. An easy way to do this is by using `kanctl`.
```

> # Please make sure the value of blueprint matches with the name of blueprint that was created earlier

Reviewer: ```suggestion
# Make sure the value of blueprint matches the name of blueprint created earlier
```
Second inconsistency noticed here and in other docs as well. We should either use `Please` everywhere or not use at all.

# --pvc argument specifies the PVC on which the action is performed
$ kanctl create actionset --action backup --namespace kanister --blueprint mysql-csi-snapshot-bp --pvc mysql/$(kubectl get pvc -n mysql --selector=app.kubernetes.io/instance=mysql-release -o=jsonpath='{.items[0].metadata.name}')
actionset backup-mlvcv created

$ kubectl --namespace kanister get actionset
NAME                         AGE
backup-mlvcv                 112s

# View the status of the actionset

> # Please make sure the name of the actionset here matches with name of the actionset created above

Reviewer: ```suggestion
# Make sure the name of the actionset here matches the name of the actionset created above
```

# View the status of the actionset
# Please make sure the name of the actionset here matches with name of the name of actionset that we have created already
$ kubectl --namespace kanister describe actionset backup-mlvcv

# Check the CSI VolumeSnapshot created
$ kubectl -n mysql get volumesnapshot
NAME                                  READYTOUSE   SOURCEPVC              SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS      SNAPSHOTCONTENT                                    CREATIONTIME   AGE

> data-mysql-release-0-snapshot-cd26z   true         data-mysql-release-0                           8Gi           do-block-storage   snapcontent-2e411d1d-0b5f-48d4-9b79-b19e824c2e38   3h40m          3h40m

Reviewer 2: Mention that you are using DO cluster hence the snapshhotclass is `do-block-storage`

## Disaster strikes!

Let's say someone accidentally deleted the test database.

```bash
# Connect to MySQL by running a shell inside MySQL's pod
$ kubectl exec -it $(kubectl get pods -n mysql --selector=app.kubernetes.io/instance=mysql-release -o=jsonpath='{.items[0].metadata.name}') -n mysql -- bash

$ mysql --user=root --password=asd#45@mysqlEXAMPLE

mysql> SHOW DATABASES;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| my_database        |
| mysql              |
| performance_schema |
| sys                |
| test               |
+--------------------+
6 rows in set (0.00 sec)

mysql> DROP DATABASE test;
Query OK, 1 row affected (0.12 sec)

mysql> SHOW DATABASES;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| my_database        |
| mysql              |
| performance_schema |
| sys                |
+--------------------+
5 rows in set (0.01 sec)
```

## Restore MySQL DB

> To restore the missing data, you should use the backup that you created before. We will use `kanctl` for creating this restore action.

Reviewer: ```suggestion
To restore the missing data, use the backup created earlier. This can be achieved using `kanctl` for creating the restore action.
```
Noticed this inconsistency in all the other docs as well. Something to do in the future for others but let's start with this one. We should either not use `you, your etc.` or `we, us, our etc.` anywhere in the document or use it everywhere. I would recommend making the documentation neutral.
For example:
One of the sentences above says
`Create the Blueprint in the kanister namespace.` which is neutral instead of saying, `You can create a blueprint in the kanister namespace` which is directed towards the reader.

```bash
# Make sure to use correct backup actionset name here
$ kanctl --namespace kanister create actionset --action restore --from backup-mlvcv
actionset restore-backup-mlvcv-6z9xn created

> # Check the PVCs in mysql namespace. You should see a new PVC in the format '<PVC-name>-restored'

Reviewer: ```suggestion
# Check the PVCs in mysql namespace. A new PVC is created with the name of format '<PVC-name>-restored'
```

$ kubectl -n mysql  get pvc
NAME                            STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS       AGE
data-mysql-release-0            Bound    pvc-b966c5ef-3d20-4ec2-9ab3-1a868737ea0a   8Gi        RWO            do-block-storage   4m14s
data-mysql-release-0-restored   Bound    pvc-2eb6fb89-78bc-4c2e-8c1e-c270d664f311   8Gi        RWO            do-block-storage   18s

> # Use following command to debug if a new PVC in the format '<PVC-name>-restored' is not created

Reviewer: ```suggestion
# In case a new PVC is not created, use the following command to check events in the actionset
```

$ kubectl --namespace kanister describe actionset restore-backup-mlvcv-6z9xn
```

> 

Reviewer: Not very confident with this approach. Here's what I think
1. Somewhere above, we should probably add a note, saying that this process may not work if application-consistent backups are required.
2. Recommend scaling down the workload (SS) deleting the existing PVC and restoring one with the same name above, and finally scaling up the workload to achieve a consistent restore.

Next, we need to add the newly restored PVC `data-mysql-release-0-restored` in the `mysql-release` StatefulSet spec template. One of the easiest way you can do that is by using `kubectl patch` command.

```bash
$ kubectl -n mysql patch statefulset mysql-release --type=json -p='[{"op": "add", "path": "/spec/template/spec/volumes/-", "value": {"name": "restored-data", "persistentVolumeClaim": {"claimName": "data-mysql-release-0-restored"}}}]'

$ kubectl -n mysql patch statefulset mysql-release --type=json -p='[{"op": "add", "path": "/spec/template/spec/containers/0/volumeMounts", "value": [{"mountPath": "/bitnami/mysql", "name": "restored-data"}, {"mountPath": "/opt/bitnami/mysql/conf/my.cnf", "name": "config", "subPath": "my.cnf"}]}]'
```

Once the patch is complete and MySQL pod is set to “Running“, you can check if the data has been restored to previous state.

## Verify the restored application data

> To verify if restore was successful, you need to connect to the MySQL CLI and query the `test` database that was created in [this](https://github.com/kanisterio/kanister/tree/master/examples/stable/mysql-csi-snapshot#create-application-data) step.

Reviewer: ```suggestion
To verify if the restore was successful, connect to the MySQL CLI and query the `test` database that was created in [this](https://github.com/kanisterio/kanister/tree/master/examples/stable/mysql-csi-snapshot#create-application-data) step.
```

```bash
# Enter into a shell inside MySQL's pod
$ kubectl exec -it $(kubectl get pods -n mysql --selector=app.kubernetes.io/instance=mysql-release -o=jsonpath='{.items[0].metadata.name}') -n mysql -- bash

# Login into the MySQL CLI
$ mysql --user=root --password=asd#45@mysqlEXAMPLE

mysql> SHOW DATABASES;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| my_database        |
| mysql              |
| performance_schema |
| sys                |
| test               |
+--------------------+
6 rows in set (0.01 sec)

mysql> USE test;
Database changed

# View data in "employees" table
mysql> SELECT * FROM employees;
+------+---------+-------------------------------+------+------------+-------------+
| name | manager | address                       | sex  | birth      | department  |
+------+---------+-------------------------------+------+------------+-------------+
| Max  | Tom     | Naperville, Illinois, Chicago | m    | 1999-03-30 | Engineering |
+------+---------+-------------------------------+------+------------+-------------+
1 row in set (0.00 sec)
```

> Thus, with the help of Kanister, we have successfully backed up our MySQL application and recovered it after a disaster.

Reviewer: ```suggestion
```
Not required

## Delete the Artifacts

The CSI VolumeSnapshot created by the backup action can be cleaned up using the following command.

```bash
# Make sure to use correct backup actionset name here
$ kanctl --namespace kanister create actionset --action delete --from backup-mlvcv
actionset delete-backup-mlvcv-cq6bw created

# View the status of the ActionSet
$ kubectl --namespace kanister describe actionset delete-backup-glptq-cq6bw
```

## Cleanup

### Uninstalling the helm chart

To uninstall/delete the `mysql-release` deployment.

```bash

> # Helm Version 3

Reviewer: ```suggestion
# Delete the MySQL database
```

# Backup and restore MySQL DB using CSI VolumeSnapshot

VolumeSnapshots provide Kubernetes users with a standardized way to copy a volume's contents at a particular point in time without creating an entirely new volume. This functionality enables, for example, database administrators to backup databases before performing edit or delete modifications.

## Introduction

> This document explains how Kanister leverages the use of CSI VolumeSnapshots to take backup and restore of databases in MySQL.

Reviewer 4: ```suggestion
This document explains how Kanister leverages the use of CSI VolumeSnapshots to snapshot and restore MySQL database.
```

Author: How about saying "....CSI VolumeSnapshots to **capture** and restore MySQL database"? Using snapshot again doesn't sound good when you read it.

## Prerequisites

- Helm 3 installed
- Kubernetes 1.16+ with Beta APIs enabled
- PV provisioner support in the underlying infrastructure
- Kanister controller version 0.71.0 installed in your cluster, let's assume in Namespace `kanister`
- Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)
- CSI Driver installed
- VolumeSnapshotClass and StorageClass resources defined (uses above CSI Driver)

> - Snapshot APIs enabled

Reviewer 4: I think all these and third point  can be combined into
- VolumeSnapshots should be supported by CSI Driver and respective CRDs should be available

```bash
$ mysql --user=root --password=asd#45@mysqlEXAMPLE

mysql> CREATE DATABASE test;
Query OK, 1 row affected (0.00 sec)

mysql> USE test;
Database changed

# Create "employees" table
mysql> CREATE TABLE employees (name VARCHAR(20), manager VARCHAR(20), address VARCHAR(200), sex CHAR(1), birth DATE, department VARCHAR(20));
Query OK, 0 rows affected (0.02 sec)

# Insert row to the table
mysql> INSERT INTO employees VALUES ('Max','Tom','Naperville, Illinois, Chicago','m','1999-03-30','Engineering');
Query OK, 1 row affected (0.01 sec)

# View data in "employees" table
mysql> SELECT * FROM employees;
+------+---------+-------------------------------+------+------------+-------------+
| name | manager | address                       | sex  | birth      | department  |
+------+---------+-------------------------------+------+------------+-------------+
| Max  | Tom     | Naperville, Illinois, Chicago | m    | 1999-03-30 | Engineering |
+------+---------+-------------------------------+------+------------+-------------+
1 row in set (0.00 sec)
```

## Backup MySQL DB

### Create Blueprint

Create Blueprint in the same namespace as the Kanister controller.

> > **Note**: We used a Kubernetes cluster on DigitalOcean. Hence, snapshotClass and storageClass in the ./mysql-csi-snapshot-blueprint.yaml file is set to `do-block-storage`. Please correct these arguments as per your cluster setup. You can make use of `kubectl patch` or `kubectl edit` commands to make these changes.

Reviewer 4: ```suggestion
> **Note**: We have set up the Kubernetes cluster on DigitalOcean. Hence, `snapshotClass` and `storageClass` in the `./mysql-csi-snapshot-blueprint.yaml` file is set to `do-block-storage`. Please correct these arguments as per your cluster setup before creating the blueprint.
```

```bash
$ kubectl create -f ./mysql-csi-snapshot-blueprint.yaml -n kanister
```

### Backup the application data

> Take a backup of the MySQL data using the backup ActionSet from above blueprint. Create an ActionSet in the `kanister` namespace. An easy way to do this is by using `kanctl`.

Reviewer 4: ```suggestion
Snapshot the MySQL database using the backup ActionSet from above blueprint. Create an ActionSet in the `kanister` namespace. An easy way to do this is by using `kanctl`.
```

```bash
# Create Actionset

> # Please make sure the value of blueprint matches with the name of blueprint that was created earlier

Reviewer 4: let's explain `--pvc` thing here.

Author: Done

data-mysql-release-0-snapshot-cd26z   true         data-mysql-release-0                           8Gi           do-block-storage   snapcontent-2e411d1d-0b5f-48d4-9b79-b19e824c2e38   3h40m          3h40m
```

## Disaster strikes!

Let's say someone accidentally deleted the test database.

```bash
# Connect to MySQL by running a shell inside MySQL's pod
$ kubectl exec -it $(kubectl get pods -n mysql --selector=app.kubernetes.io/instance=mysql-release -o=jsonpath='{.items[0].metadata.name}') -n mysql -- bash

$ mysql --user=root --password=asd#45@mysqlEXAMPLE

mysql> SHOW DATABASES;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| my_database        |
| mysql              |
| performance_schema |
| sys                |
| test               |
+--------------------+
6 rows in set (0.00 sec)

mysql> DROP DATABASE test;
Query OK, 1 row affected (0.12 sec)

mysql> SHOW DATABASES;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| my_database        |
| mysql              |
| performance_schema |
| sys                |
+--------------------+
5 rows in set (0.01 sec)
```

## Restore MySQL DB

> To restore the missing data, you should use the backup that you created before. We use `kanctl` for creating this restore action. `kanctl` helps create ActionSets that depend on other ActionSets.

Reviewer 4: ```suggestion
To restore the missing data, you should use the backup that you created before. We will use `kanctl` for creating this restore action.
```

actionset restore-backup-mlvcv-6z9xn created

# Check the PVCs in mysql namespace. You should see a new PVC in the format '<PVC-name>-restored'
$ kubectl -n mysql  get pvc
NAME                            STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS       AGE
data-mysql-release-0            Bound    pvc-b966c5ef-3d20-4ec2-9ab3-1a868737ea0a   8Gi        RWO            do-block-storage   4m14s
data-mysql-release-0-restored   Bound    pvc-2eb6fb89-78bc-4c2e-8c1e-c270d664f311   8Gi        RWO            do-block-storage   18s

# Use following command to debug if a new PVC in the format '<PVC-name>-restored' is not created
$ kubectl --namespace kanister describe actionset restore-backup-mlvcv-6z9xn
```

Next, we need to add the newly restored PVC `data-mysql-release-0-restored` in the `mysql-release` StatefulSet spec template. One of the easiest way you can do that is by using `kubectl patch` command.

```bash
$ kubectl -n mysql patch statefulset mysql-release --type=json -p='[{"op": "add", "path": "/spec/template/spec/volumes/-", "value": {"name": "restored-data", "persistentVolumeClaim": {"claimName": "data-mysql-release-0-restored"}}}]'

> $ kubectl -n mysql patch statefulset mysql-release --type=json -p='[{"op": "add", "path": "/spec/template/spec/containers/0/volumeMounts", "value": [{"mountPath": "/bitnami/mysql", "name": "restored-data"}, {"mountPath": "/opt/bitnami/mysql/conf/my.cnf", "name": "config", "subPath": "my.cnf"}]}]'

Reviewer 4: should these be statefulsets or deployments.

Author: These should be statefulsets

```

Once the patch is complete and MySQL pod is set to “Running“, you can check if the data has been restored to previous state.

## Verify the restored application data

> To verify if restore was successful, you need to connect to the MySQL CLI and query the test database that was created in [this](https://github.com/kanisterio/kanister/tree/master/examples/stable/mysql-csi-snapshot#create-application-data) step.

Reviewer 4: ```suggestion
To verify if restore was successful, you need to connect to the MySQL CLI and query the `test` database that was created in [this](https://github.com/kanisterio/kanister/tree/master/examples/stable/mysql-csi-snapshot#create-application-data) step.
```

```bash
# Enter into a shell inside MySQL's pod
$ kubectl exec -it $(kubectl get pods -n mysql --selector=app.kubernetes.io/instance=mysql-release -o=jsonpath='{.items[0].metadata.name}') -n mysql -- bash

# Login into the MySQL CLI
$ mysql --user=root --password=asd#45@mysqlEXAMPLE

mysql> SHOW DATABASES;
+--------------------+
| Database           |
+--------------------+
| information_schema |
| my_database        |
| mysql              |
| performance_schema |
| sys                |
| test               |
+--------------------+
6 rows in set (0.01 sec)

mysql> USE test;
Database changed

# View data in "employees" table
mysql> SELECT * FROM employees;
+------+---------+-------------------------------+------+------------+-------------+
| name | manager | address                       | sex  | birth      | department  |
+------+---------+-------------------------------+------+------------+-------------+
| Max  | Tom     | Naperville, Illinois, Chicago | m    | 1999-03-30 | Engineering |
+------+---------+-------------------------------+------+------------+-------------+
1 row in set (0.00 sec)
```

> Thus, with the help of Kanister, we have successfully backed up our MySQL application data and recovered it after a disaster.

Reviewer 4: ```suggestion
Thus, with the help of Kanister, we have successfully backed up our MySQL application and recovered it after a disaster.
```
